### ISSUE TITLE
install-packages caching fails on Windows

### ISSUE THREAD

**Issue by Peter-Darton-i2** [CircleCI-Public/python-orb#123]
## Orb version:

```yaml
orbs:
  python: circleci/python@2.2.0
```

## What happened:

I'm writing a CircleCI build file that packages a python project on both Windows and Linux.
I have a job with a parameterise executor that calls `checkout` and then calls the `install-packages` command.

On [Linux](https://app.circleci.com/pipelines/github/i2group/ias-client/98/workflows/bf811bcd-6c0a-41e3-a73e-e8a208ccb4c7/jobs/191), the `install-packages` command's use of `restore_cache` & `save_cache` works fine:
![image](https://github.com/user-attachments/assets/37bae5df-5245-46fb-b750-e36c6da6b92b)

On [Windows](https://app.circleci.com/pipelines/github/i2group/ias-client/98/workflows/bf811bcd-6c0a-41e3-a73e-e8a208ccb4c7/jobs/192), these steps are red with an error:
![image](https://github.com/user-attachments/assets/d37b09ad-2230-4491-878a-429edce6c611)
The "Restoring cache" step errors saying `error computing cache key: template: cacheKey:1:18: executing "cacheKey" at <checksum "/tmp/cci_pycache/lockfile">: error calling checksum: open C:\tmp\cci_pycache\lockfile: The system cannot find the path specified.`
The Saving cache" step errors saying `Error computing cache key: template: cacheKey:1:18: executing "cacheKey" at <checksum "/tmp/cci_pycache/lockfile">: error calling checksum: open C:\tmp\cci_pycache\lockfile: The system cannot find the path specified.`
... but the job continues, albeit without any caching functionality.

This lack of caching results in bad performance - it took 8 seconds on Linux (7 seconds to restore the cache and 1 to use it) but 49 seconds on Windows (45 seconds to re-download everything, 4 to use it)

## Expected behavior:

Caching should work (perfectly) on all OSs: CircleCI orbs should _not_ be Linux-specific.
While I would expect things might take longer on Windows than Linux, the Microsoft-performance-penalty is usually only "takes twice as long compared to Linux" whereas the lack of caching is making it take take *six* times as long ... and results in way more (reliance upon) network traffic (and hence build instability).

Also, ideally, I shouldn't have to switch the Windows agent's default shell to bash either.
As-is, the CircleCI Windows orb defaults all CircleCI Windows agents to use a default shell of Powershell, so either all `run` commands then have to specify bash, or users have to override the default when specifying the executor (which is what I've done here) ... or the Windows orb needs to change to default to bash (which would be a major breaking change, albeit maybe a welcome one).
So it would be nice if the python orb's bash `run` commands declared that they're bash commands so that a Windows machine doesn't try to run them as Powershell.

## Additional Information:

Heavily-trimmed CircleCI config.yml:
```yaml
version: 2.1

orbs:
  python: circleci/python@2
  windows: circleci/windows@5

executors:
  linux:
    docker:
      - image: cimg/python:3.12
    resource_class: small

commands:
  python-install-pipenv-packages:
    parameters:
      app-dir:
        type: string
    steps:
      - python/install-packages:
          app-dir: <<parameters.app-dir>>
          pkg-manager: pipenv
          include-branch-in-cache-key: false
          args: --dev
  install-python-on-windows:
    steps:
      - run:
          name: Install Python 3.12
          command: |
            set -euxo pipefail
            choco feature enable -n allowGlobalConfirmation
            # skip python install if we already have something that's good enough
            if ! python --version | grep -q 'Python 3\.12'; then
              choco upgrade python312 --package-parameters="/NoLockdown"
            fi
            pip install pipenv

jobs:
  python-cache-dependencies:
    parameters:
      app-dir:
        type: string
      executor:
        type: executor
    executor: << parameters.executor >>
    steps:
      - checkout
      - python-install-pipenv-packages:
          app-dir: <<parameters.app-dir>>

workflows:
  build-and-test:
    jobs:
      - python-cache-dependencies:
          name: server cache dependencies linux
          app-dir: server
          executor: linux
      - python-cache-dependencies:
          name: server cache dependencies windows
          app-dir: server
          executor:
            name: windows/default
            shell: bash.exe
          pre-steps:
            - install-python-on-windows
```

Note: my GitHub repo contains a `server/` folder that contains a `Pipfile` and `Pipfile.lock` file (as well as the rest of my python code), hence the need for `app-dir: server` everywhere.

I think the fault arises because of a discrepancy between where a Windows bash command thinks `/tmp` is located within the Windows OS' filesystem, and where the CircleCI code handling the build instructions thinks `/tmp` is within the Windows OS' filesystem.

**Comment by Peter-Darton-i2** [comment]
## Possible resolution methods

I found that if I asked my CircleCI build config.yml to do e.g. `store_artifacts` from `/tmp/foo` on a Windows machine then it'd complain it couldn't find `C:\tmp\foo` ... even though a previous bash step had populated `/tmp/foo`.
The solution that I used there was to stop using an absolute path (I was also using /tmp for my temporary file!) and instead tell bash to make a `.build/foo` folder and then to point `store_artifacts` at `.build/foo`.

So I can think of 3 potential solutions...

1. Use a relative path
   I suspect that if the python orb were to calculate the checksum using a file `.cci_pycache_lockfile` instead of `/tmp/cci_pycache/lockfile` then both the `restore_cache` (etc) and the bash code (that created the file) would agree on where the file was on the Filesystem and things would "just work".
1. Generate the file into a C:\tmp when on Windows.
   Make the script which creates the `lockfile` file go and see "if on Windows then write into `C:\tmp` instead of `/tmp`. That way the file would be where the config.yml expects it to be.
1. Change the CircleCI agent code on Windows
   Get the CircleCI agent code to do the bash <-> Windows translation so that when it was told to checksum a file in /tmp/foo it looked at where _bash_ would put /tmp/foo instead of just C:\tmp - it's probably easier to make the orb OS-agnostic...)
   ... but probably best not do this if someone's using `\` path separators - probably best to only to a "translate into bash path" for unix-style paths.

**Comment by marboledacci** [comment]
Hi @Peter-Darton-i2, I like your first idea about using a relative path for the files. I'm doing some tests to see how it goes.
About the bash and windows, the problem with this is that all the options lead to a point where you might have to manually specify the shell and the less problematic is the current one, where you set bash in the executor

**Comment by Peter-Darton-i2** [comment]
FYI I hit pretty much this exact same issue when writing a command that created a tar which it then uploaded as a CircleCI artifact (and/or as a workspace file).
I solved the problem by stopping using `/tmp/`... as the base folder in my bash code and used `.circleci/.tmp/`... instead - the "works on Linux but not on Windows" issues went away.
So, yes, I believe option (1) will work.

Note: I also hit "resource busy" errors when deleting files on Windows because, having moved my files out of /tmp into the local folder, I wanted to tidy up the mess before I moved onto further build steps. I have no idea what process had locked the file(s) open, but something had rendered them non-deletable causing `rm -rf `... to fail ... even when retried multiple times over several minutes.
i.e. beware that you might not be able to "clean up" any mess you make on Windows, so make sure that any mess is unlikely to cause an issue if left there.

PS. Personally I think option (3) would be the nicest option as then the problem would simply "go away" for all orbs, not just this one ... but I appreciate that that's an RFE for another CircleCI team...
